### PR TITLE
vulkan / amdgpu-pro update

### PR DESCRIPTION
**Pull request by corngood**
## vulkan
- Leftover change by @Ralith from the previous PR (not sure what happened there)
- Search XDG paths for layers/drivers (as suggested by @Ralith)
  - Uses an unfortunately large patch to the loader
  - Need to figure out how to upstream this
- Output validation layers and demos
  - In the future we should probably move headers to .dev, layers to .layers (the layers just need rpath fixes for that)
## amdgpu-pro
- Install icd json to /share/vulkan
- Fixes for linux-4.8 (tested HAWAII on 4.8.3)
  - Pulls shared patches from gentoo vulkan-overlay repository to make that relationship explicit
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

@expipiplus1 @vcunat @Ralith  - Thoughts?

### CODE REVIEW

File: pkgs/development/libraries/vulkan-loader/default.nix [hunk 1]
```diff
@@ -1,8 +1,6 @@
 { stdenv, fetchgit, fetchFromGitHub, cmake, pkgconfig, git, python3,
   python3Packages, glslang, spirv-tools, x11, libxcb, wayland }:
 
-assert stdenv.system == "x86_64-linux";
```
Comment (expipiplus1) [inline pkgs/development/libraries/vulkan-loader/default.nix]: Working on non-linux platforms now?

**Comment by expipiplus1** [comment]
Looks good!

**Comment by Ralith** [comment]
The loader needs to be further patched to look in `/run/opengl-driver/` or `/run/opengl-driver-32` (depending on target arch) for the ICD json and library, and probably for all the other types of components as well. In https://github.com/NixOS/nixpkgs/pull/19920 I accomplished this by replacing all occurrences of `"/usr/"` with the architecture-appropriate driver path in `vk_loader_platform.h`.

Have you had any success actually running a vulkan rendering demo yet?

**Comment by corngood** [comment]
@Ralith I've only tested on this amdgpu-pro driver, but it now installs in system packages, so I have
`/run/current-system/sw/share/vulkan/icd.d/amd_icd64.json` which points to `/nix/store/avrgqn78r2z7286sf76hj7d0fl5ndndb-amdgpu-pro-16.30-4.8.3/lib/amdvlk64.so`.

The layers are currently still in /share of the vulkan-loader package.  Since It searches XDG_DATA_DIRS, the loader and layers should work in either the system or user environment.

Nothing should really be needed in LD_LIBRARY_PATH or /etc.

The sdk demos work with and without validation.  You can build them with `nix-build -A vulkan-loader.demos`.

I've been trying to package something non-trivial using vk, but I haven't found a great candidate yet.  vkQuake builds and runs for me, but it needs some more work to package the non-free demo data (similar to the quake3 packages) so you can play it without manually unpacking the data.  The dolphin emulator hasn't had a release since the vk renderer went in.  Suggestions are welcome.

**Comment by Ralith** [comment]
@corngood That will break if you upgrade your kernel module to a version with a different interface. Compare https://github.com/NixOS/nixpkgs/pull/19918 which installs the ICD in the correct location. `/run/opengl-driver` is *not* in `XDG_DATA_DIRS`, and so must be compiled into the loader. It's conceivable that ICDs might come with layers that interact with the ICD's internals or the kernel, so looking for layers and so forth there too might also be prudent.

**Comment by Ralith** [comment]
Er, rather, your current approach break if you upgrade your driver to one that expects a different kernel module interface without rebooting.

The good news is, after hacking `XDG_DATA_DIRS`, the demos actually run for me with this build of the loader and #19918 installed, unlike under #19920. We're close!

**Comment by corngood** [comment]
Ah, yeah, you could definitely be broken before rebooting.  I'm not sure about the solution to that though.  You could have multiple ICD packages that use `boot.extraModulePackages` and `environment.systemPackages`, but only one can be `boot.opengl.package`.

**Comment by Ralith** [comment]
You mean `hardware.opengl`? Perhaps ICDs could be most idiomatically placed in `hardware.opengl.extraPackages`.

**Comment by corngood** [comment]
Yeah, that's what I meant.  I always figured `/run/opengl-driver` was a bit of a hack that could be retired if something like glvnd was supported by all drivers.  I'm also a bit fuzzy on this, but a similar situation must exist with other ICDs like vdpau, opencl.

Maybe @vcunat can comment?

**Comment by Ralith** [comment]
`hardware.opengl.extraPackages` is what's currently used for vdpau and vaapi, it looks like.

**Comment by corngood** [comment]
Updated amdgpu-pro to 16.40, which was just released.  It seems much more stable on my 290:

- no IOMMU page faults at boot
- no errors in vulkan smoke test
- no freezing when using vdpau

**Comment by vcunat** [comment]
`/run/opengl-driver{,-32}` seems unlikely to go away anytime soon, even if we deploy glvnd (which isn't supported by fglrx nor amdgpu now and may have nontrivial performance impact). For example, vdpau and vaapi most likely need it to locate some libs, via `mesa_noglu.driverLink` path.

File: pkgs/development/libraries/vulkan-loader/default.nix (hunk 1 above)
Comment (vcunat) [inline pkgs/development/libraries/vulkan-loader/default.nix]: `meta.platforms` list is probably the best place. Only three system values are commonly used, I think, so the rest seems not too much significant anyway.

**Comment by corngood** [comment]
@vcunat So what do you think about deploying the ICD configs to `/run/current-system/sw/share/vulkan/icd.d/amd_icd64.json` and having the loader look in there?  That's the way this currently works, but @Ralith had some concerns.

Basically, if we can make vulkan work without /run/opengl-driver, should we?

**Comment by vcunat** [comment]
`/run/booted-system/...` would be a closer shot, I expect. I currently don't see a significant advantage for either (`/run/opengl-driver` vs. `/run/booted-system`), but honestly I don't even know what ICDs are for, so I don't seem a good judge for that.

**Comment by Ralith** [comment]
The ICD, or installable client driver, is the userspace component of the GPU driver. It needs to be kept in sync with the kernel component. The configs are json metadata which direct the loader to the correct shared library.

**Comment by Ralith** [comment]
@vcunat I think searching `/run/booted-system` does not make sense, because that contains the complete system, so if you e.g. update a normal layer, both the updated and old versions would show up in the search, whereas only showing the updated version would be correct. Only components of the driver should be pinned to the booted system, and that's what `/run/opengl-driver` accomplishes.

**Comment by corngood** [comment]
It looks like `/run/opengl-driver` is not actually pinned to the booted system.  It's linked in an activation script, which happens on system switch.  I also noticed, as an example, `/sys/module/firmware_class/parameters` gets updated on system switch to point to the new firmwares.

I feel like there are multiple levels at which problems could occur when updating, whether the icd is loaded from booted-system, current-system, or opengl-driver.

Just to be clear, I still think looking in all the XDG_CONFIG, and XDG_DATA dirs is a good idea for both layers and ICDs.  Layers, and ICDs without kernel dependencies could easily be installed either in the system or user environment.  Nobody has a problem with that part, right?

The question is should there be additional search paths, and if so, what they should be?

FWIW, the user space driver is not strictly coupled to the kernel module.  The AMDGPU-PRO user space will work reasonably well with linux-4.7 amdgpu, there are just some problems with fences in vulkan, etc.  It seems to be aware of the kernel capabilities and will print a warning at xorg startup.

**Comment by Ralith** [comment]
> It looks like /run/opengl-driver is not actually pinned to the booted system. It's linked in an activation script, which happens on system switch.

That can't be right. I'm pretty sure existing proprietary drivers rely on being in sync. Can you share where you saw that?

> I also noticed, as an example, /sys/module/firmware_class/parameters gets updated on system switch to point to the new firmwares.

Doesn't that only get read on boot anyway?

> I feel like there are multiple levels at which problems could occur when updating, whether the icd is loaded from booted-system, current-system, or opengl-driver.

Loading the ICD *alone* from booted-system should be fine, I think. I speculated that it might be nice to have support for layers that rely on internal details of the kernel module or ICD, which would require a dedicated path for ICD/kernel module sensitive things in general vs. just patching the loader to look in one more place for the ICD alone, but as far as I know no such layers currently exist, so perhaps it should not be a priority. If the ICD is loaded from the booted system and everything else is loaded from the current system, I don't see any problems. Am I missing something?

>  Nobody has a problem with [searching XDG paths], right?

Right, that part's obviously correct, IMO, to the point that it probably can and should be upstreamed, as you suggested in the PR.

> FWIW, the user space driver is not strictly coupled to the kernel module.

My understanding is that this is not usually the case, though perhaps I'm wrong. Certainly, if userspace components *can* be tightly bound to a kernel module, then we must account for that for `nixos-rebuild switch` to be reliable.

**Comment by vcunat** [comment]
`nixos-rebuild switch` can be a bit fragile on significant changes, as in many cases "we" aren't careful enough (sometimes there might be upstream packages to blame).

In this particular case, the `/run/opengl-drivers{,32}` symlinks are [setup by](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/hardware/opengl.nix#L123) `system.activationScripts`... which is run immediately on `switch`, so it should go badly at least if you switch vendor (like `"nvidia"` <-> `"nouveau"`).

IIRC it used to be done when starting X, so it "only" went bad after restarting it (not done by default)... but I believe in general it isn't possible to always do the opengl switch, because kernel modules are required and major kernel versions may not match.

**Comment by Ralith** [comment]
Well, that's unfortunate. It might still be best to use `/run/opengl-driver` so that if this is fixed in the future, the vulkan infrastructure benefits as well. Unless there's a good reason why it isn't pinned to the booted system?

It seems like there's a general need for a mechanism for certain packages to always be exclusively visible from the booted system, and if `/run/opengl-driver` is not that mechanism, then one should be created. I don't have a great sense of how to go about that, unfortunately.

In any case, this appears to be a fundamental problem with NixOS, and IMO shouldn't block progress on vulkan support. I'm still in favor of using `/run/opengl-driver` for kernel-sensitive ICDs and related components in the interest of making this issue easier to fix in the future. It's also consistent with the already-merged intel vulkan ICD. As it stands, this PR will produce a loader that does not work on intel systems without manually clobbering environment variables.

Perhaps the easiest fix to this PR would be to add `/run/opengl-driver{,32}/share` to `XDG_DATA_DIRS`? Not sure what the right place is for that. Perhaps use `environment.extraInit` to prefix it to the existing value?

**Comment by vcunat** [comment]
> Unless there's a good reason why it isn't pinned to the booted system?

It doesn't seem intentional (used to be on X start until 852c270035f). In some cases it seems safe to switch `/run/opengl-driver{,32}` but I'd prefer to only do it on reboot, at least until we have a more reliable heuristics when it's safe to do. I hope I'll find time soon to fix this.

**Comment by corngood** [comment]
Adding `opengl-driver[-32]` to the front of `XDG_DATA_DIRS` and `XDG_CONFIG_DIRS` seems like a good next step to me.  It feels coherent with their place in LD_LIBRARY_PATH.  I can then take the amdgpu-pro driver out of `systemPackages`.  Any concerns about that?

I do feel like `opengl-driver` is a poor name for 'things pinned to the booted kernel', but that can be discussed separately.

**Comment by corngood** [comment]
@Ralith I did as you suggested to set XDG_DATA_DIRS.

I also tested this with the bleeding edge mesa radeon icd, and finds it ok, so the intel one should work as well.  I don't have hardware here to test intel or nvidia.

I'll put my mesa/llvm changes in another PR, so people can opt-in to the open source vulkan driver.

**Comment by Ralith** [comment]
Confirmed working with the intel drivers.

It's a bit weird how having the vulkan loader installed in your user environment leads to it finding its layers twice, once from your environment and once from its hardcoded install paths. Perhaps, as a special case the loader's own layers should not be installed elsewhere, since they'll always be found directly? Is that even possible in nix? It's certainly desirable that layers normally get installed in the environment.

**Comment by vcunat** [comment]
Let's push now and re-iterate later.

**Comment by corngood** [comment]
@Ralith I'd like to move the layers to vulkan-loader.layers, I just need to deal with some rpath drama.  Also the headers should probably be in $dev.

Either way I think it would be best to disable searching the loader install path.

**Pull request merged**